addons/payment_alipay/controllers/main.py

```diff
@@ -14,54 +14,33 @@
-            if resp in ['TRADE_FINISHED', 'TRADE_SUCCESS']:
-                _logger.info('Alipay: validated data')
-            elif resp == 'TRADE_CLOSED':
-                _logger.warning('Alipay: payment refunded to user and closed the transaction')
-            else:
-                _logger.warning('Alipay: unrecognized alipay answer, received %s instead of TRADE_FINISHED/TRADE_SUCCESS and TRADE_CLOSED' % (post['trade_status']))
-        if post.get('out_trade_no') and post.get('trade_no'):
-            # TODO ARJ wtf? do we really need to do this?
-            post['reference'] = request.env['payment.transaction'].sudo().search([('reference', '=', post['out_trade_no'])]).reference
-            return request.env['payment.transaction'].sudo()._handle_feedback_data('alipay', post)
-        return False  # TODO ARJ useless rtype bool, could be None
+    _return_url = '/payment/alipay/return'  # If these are constant, they should be modified in all modules
 
     def _alipay_validate_notification(self, **post):
-        if post.get('out_trade_no'):
-            # TODO ARJ use _get_tx_from_feedback_data
-            alipay = request.env['payment.transaction'].sudo().search([('reference', '=', post.get('out_trade_no'))]).acquirer_id
-        else:
-            # TODO ARJ nope. There could be several payment.acquirer sharing the same provider (one per company for ex)
-            alipay = request.env['payment.acquirer'].sudo().search([('provider', '=', 'alipay')])
+        alipay = request.env['payment.acquirer'].sudo().search([('provider', '=', 'alipay')], limit=1)
         val = {
             'service': 'notify_verify',
             'partner': alipay.alipay_merchant_partner_id,
             'notify_id': post['notify_id']
         }
-        # TODO ARJ Do you know why we need to make this request? I could use a comment here
-        response = requests.post(alipay.alipay_get_form_action_url(), val)
+        # https://global.alipay.com/docs/ac/hk_auto_debit/notif
+        response = requests.post(alipay._get_alipay_urls(), val)
         response.raise_for_status()
         _logger.info('Validate alipay Notification %s' % response.text)
         # After program is executed, the page must print “success” (without quote). If not, Alipay server would
         # keep re-sending notification, until over 24 hour 22 minutes
         # Generally, there are 8 notifications within 25 hours (Frequency: 2m,10m,15m,1h,2h,6h,15h).
         if response.text == 'true':
-            self._alipay_validate_data(**post)
+            request.env['payment.transaction'].sudo()._handle_feedback_data('alipay', post)
             return 'success'
-        return ""  # TODO ARJ nitpicking: returning "fail" would still be a 200 but would clearly state that the notif couldn't be processed
+        return "fail"
 
     # TODO ARJ there is still the not so likely eventuality of an HTTP 307 but I think that we can safely drop the POST
```

Comment:
? drop the parameters? They are used to validate the transaction.
Maybe you mean only keep the GET ?

Reply:
Yes